compiler/rustc_middle/src/ty/inhabitedness/inhabited_predicate.rs

```diff
@@ -71,11 +77,25 @@ impl<'tcx> InhabitedPredicate<'tcx> {
                 match normalized_pred {
                     // We don't have more information than we started with, so consider inhabited.
                     Self::GenericType(_) => Ok(true),
-                    pred => pred.apply_inner(tcx, param_env, in_module),
+                    pred => {
+                        // A type which is cyclic when monomorphized can happen here since the
+                        // layout error would only trigger later. See e.g. `tests/ui/sized/recursive-type-2.rs`.
+                        if eval_stack.contains(&t) {
+                            return Ok(true); // Recover; this will error later.
+                        }
```

Comment:
What used to happen here? A stack overflow? IS there a test that exercises this and would repro it on nightly with `exhaustive_patterns `?

Reply:
Yep, [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=003924cba3ab88ea27d345e709e3bb79). It's in `tests/ui/sized/recursive-type-2.rs`